internal/app/api/client/agent_server.go

```diff
@@ -0,0 +1,91 @@
+package client
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jinzhu/gorm"
+
+	"soldr/internal/app/api/models"
+	"soldr/internal/app/api/storage/mem"
+	"soldr/internal/secret"
+	"soldr/internal/storage"
+	"soldr/internal/storage/mysql"
+)
+
+type AgentServerClient struct {
+	db      *gorm.DB
+	dbConns *mem.DBConnectionStorage
+	s3Conns *mem.S3ConnectionStorage
+}
+
+func NewAgentServerClient(
+	db *gorm.DB,
+	serviceDBConns *mem.DBConnectionStorage,
+	serviceS3Conns *mem.S3ConnectionStorage,
+) *AgentServerClient {
+	return &AgentServerClient{
+		db:      db,
+		dbConns: serviceDBConns,
+		s3Conns: serviceS3Conns,
+	}
+}
+
+func (c *AgentServerClient) GetDB(ctx context.Context, hash string) (*gorm.DB, error) {
+	db, err := c.dbConns.Get(hash)
+	if err != nil && err == mem.ErrNotFound {
+		var service models.Service
```

Comment:
I would suggest to wrap "slow" path (when connection is not found in the connections cache) into mutex lock to prevent multiple connections to the same DB in parallel.

Reply:
Also maybe following pattern will be a little bit more readable:
```
db, err := c.dbConns.Get(hash)
if err == nil { // fast path
  return db, nil
}
if err != mem.ErrNotFound { // strange cache error (maybe this shouldn't be the case even?)
  return nil, err
}
// slow path

```

Reply:
Agree, refactored